kernel/rros/observable.rs

```diff
@@ -0,0 +1,970 @@
+            Ok(n) => n,
+            Err(_) => {
+                return Err(kernel::Error::ENOMEM);
+            }
+        };
+        observer.free_list.push_back(nf);
+    }
+
+    let observer = match Arc::try_new(observer) {
+        Ok(o) => o,
+        Err(_) => {
+            // TODO:
+            // evl_put_element(&observable->element);
+            return Err(kernel::Error::ENOMEM);
+        }
+    };
+
+    let subscriber = unsafe { &mut (*sbr) };
+    let flags = subscriber.subscriptions.irq_lock_noguard();
+    let ret = unsafe {
+        (*(subscriber.subscriptions.locked_data().get())).try_insert(fundle, observer.clone())
+    };
+    subscriber.subscriptions.irq_unlock_noguard(flags);
+
+    match ret {
+        Ok(None) => {
+            // everything is right
+        }
+        Ok(Some(_)) => {
+            // allocate a Node, but RBTree has a node with the same key
+            // node will be free automatically
+            return Err(kernel::Error::EEXIST);
+        }
+        Err(info) => {
+            // can't allocate a Node
+            // observer will be free automatically
+            return Err(info);
+        }
+    }
+
+    let flags = observable.lock.raw_spin_lock_irqsave();
+    observable.observers.push_back(observer);
+
+    observable.oob_wait.lock.raw_spin_lock();
+    observable.writable_observers += 1;
+    observable.oob_wait.lock.raw_spin_unlock();
+    observable.lock.raw_spin_unlock_irqrestore(flags);
+    // TODO:
+    // evl_put_element(&observable->element);
+
+    Ok(0)
+}
+
+fn decrease_writability(writable_observers: &mut i32) {
+    *writable_observers -= 1;
+    if *writable_observers < 0 {
+        pr_err!("observable: negative quantity of writable_observers\n");
+    }
+}
+
+fn get_observer(observer: &mut RrosObserver) {
+    observer.refs += 1;
+}
+
+fn put_observer(observer: &mut RrosObserver) -> bool {
+    observer.refs -= 1;
+    let new_refs = observer.refs;
+    // pr_warn!
+    new_refs == 0
+}
+
+fn wake_oob_threads(observable: &mut RrosObservable) {
+    let flags = observable.oob_wait.lock.raw_spin_lock_irqsave();
+
+    observable.oob_wait.lock.raw_spin_unlock_irqrestore(flags);
+
+    unsafe {
+        rros_schedule();
+    }
+}
+
+fn notify_one_observer(
+    oob_wait: &mut RrosWaitQueue,
+    writable_observers: &mut i32,
+    observer: &mut RrosObserver,
+    tmpl_nfr: &mut RrosNotificationRecord,
+) -> bool {
+    let last_notice;
+    let mut nfr;
+
+    let flags = oob_wait.lock.raw_spin_lock_irqsave();
+    if observer.flags & RROS_NOTIFY_ONCHANGE != 0 {
+        last_notice = observer.last_notice.clone();
+        observer.last_notice.tag = tmpl_nfr.tag;
+        observer.last_notice.event = tmpl_nfr.event.clone();
+
+        if observer.flags & RROS_NOTIFY_INITIAL != 0 {
+            observer.flags &= !RROS_NOTIFY_INITIAL;
+        } else if last_notice.tag == tmpl_nfr.tag
+            && unsafe { last_notice.event.lval == tmpl_nfr.event.lval }
```

Comment:
Why use unsafe here?

Reply:
Because `event` is an `ObservableValue` type, which is a union. Both read and write operations on unions are unsafe. The reason I use a union is that `ObservableValue` is only used when interacting with the user, and the unsafe does not propagate elsewhere.

Reply:
Don't use union, and replace it with enum.